### PR TITLE
trd: license and copyright policy

### PR DESCRIPTION
**Pull request by ppannuto**
### Pull Request Overview

First draft of TRD for license and copyright rules for the Tock project.

### Testing Strategy

N/A

### TODO or Help Wanted

Review of language and policy decisions.

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [ ] Ran `make prepush`.

### CODE REVIEW

File: doc/reference/trd-legal.md
Comment (hudson-ayers) [inline doc/reference/trd-legal.md:86]: So this is slightly different than Amit's approach in https://github.com/tock/tock/pull/3317, in that it does not require a copyright notice in all files. I personally think excluding the copyright notice except when individuals want to include it sounds nice.

**Review by hudson-ayers — COMMENTED** [review]
Updating https://github.com/tock/tock/pull/3317 to match the process in this document, as best as I can tell, would require the following changes:

1. Updating non-rust files that are still code files (e.g. the python/bash code in `tools/` and our Makefiles) to include the license notice
2. Remove the "Copyright Tock Contributors" line from each file, only have a Copyright line if someone chooses to add it.
3. Add the line "Licensed under the Apache License, Version 2.0 or the MIT License." as the first line of each file, before the SPDX license identifier.

Overall this approach looks good to me. Curious to hear what @jrvanwhy in particular thinks about not requiring the "Copyright Tock Contributors" line on all files without a different copyright assignment

File: doc/reference/trd-legal.md
Comment (brghena) [inline doc/reference/trd-legal.md:98]: "exactly one" or "at least one"? I don't know if it's necessary to be this picky here.

**Review by brghena — DISMISSED** [review]
This all looks like a reasonable approach to me.

File: doc/reference/trd-legal.md
Comment (jrvanwhy) [inline doc/reference/trd-legal.md:48]: I don't think we should explain *why* a user should prefer one license over the other. IMO that gets too close to sounding like legal advice.

File: doc/reference/trd-legal.md
Comment (jrvanwhy) [inline doc/reference/trd-legal.md:62]: Was this supposed to say "Copyright notices are a common means to provide this."? "Copyright" is a much broader term than "Copyright notices".

File: doc/reference/trd-legal.md
Comment (jrvanwhy) [inline doc/reference/trd-legal.md:83]: How do we deal with files like [`chips/nrf52/src/crt1.rs`](https://github.com/tock/tock/blob/0f7fe7d9355002e4e7065cc3fb940a0d121e8d21/chips/nrf52/src/crt1.rs) that have a copyright notice in the middle of the file?

File: doc/reference/trd-legal.md
Comment (jrvanwhy) [inline doc/reference/trd-legal.md:89]: The entire "Format" section makes it difficult to re-use code from other projects, as their license and copyright information may not match our requirements, and I do not think we can arbitrarily reformat those notices.

An example of a file that does not match this format is [`capsules/src/sip_hash.rs`](https://github.com/tock/tock/blob/0f7fe7d9355002e4e7065cc3fb940a0d121e8d21/capsules/src/sip_hash.rs).

File: doc/reference/trd-legal.md
Comment (jrvanwhy) [inline doc/reference/trd-legal.md:134]: I'm not sure that "relinquish the initial copyright" is something you can *do* in many jurisdictions. Probably better to leave this half of the sentence out.

File: doc/reference/trd-legal.md
Comment (alevy) [reply doc/reference/trd-legal.md:48]: Agree. Just say less

File: doc/reference/trd-legal.md
Comment (alevy) [reply doc/reference/trd-legal.md:83]: We can move that copyright notice

**Comment by alevy** [comment]
We're going to take another pass on this to condense it and remove some unnecessary requirements.

Unlike most other PRs, let's hold off on merging after approvals in order to get feedback from external contributors first.

But please approve once you're comfortable so we _can_ be confident that what we get feedback on is similar to what we'll merge.

(draft coming later today)

File: doc/reference/trd-legal.md
Comment (phil-levis) [inline doc/reference/trd-legal.md:101]: I think we want to avoid specifying this at such a low level. E.g., the standard format of a BSD license in a comment block has lines of whitesapce between the enumerated elements. That is, remove this MUST NOT.

File: doc/reference/trd-legal.md
Comment (phil-levis) [inline doc/reference/trd-legal.md:6]: Best Common Practice?

**Comment by ppannuto** [comment]
Quick comment re the formatting bits: the pedantic, low-level bits were written with an eye towards making Johnathan's life easier re: automatic enforcement. I do think top of file is important and clustering licenses and copyright; the rest I have less attachment to.

**Comment by jrvanwhy** [comment]
> Quick comment re the formatting bits: the pedantic, low-level bits were written with an eye towards making Johnathan's life easier re: automatic enforcement. I do think top of file is important and clustering licenses and copyright; the rest I have less attachment to.

In this respect, I think there are two distinct groups of license notices:

**Tock project notices:** Notices added by Tock contributors when they contribute to the Tock project.
**3p notices:** Notices in code that the Tock project copied from another project, such as the `sip_hash.rs` notice.

When I commented on the Format section, I assumed this policy would apply to both sets of license notices. I think that being picky about 3p notices makes it difficult to import code from other projects.

We could limit our formatting requirements and automated enforcement to Tock project notices. Then we could import code from other projects as long as that code is also (Apache 2.0 OR MIT)-licensed, while remaining picky about *most* of the license headers in the repository.

We'll probably end up with an exceptions list in the tool. For example, if someone copies a file from another project without making any changes to it, then it should be committed without a Tock project notice. In that case, we'll need to list the file as an exception in the tool. When we eventually make our own contributions to that file, we will then add a Tock project notice and remove the license exception.

**Comment by jrvanwhy** [comment]
[Rendered view](https://github.com/tock/tock/blob/doc-copy-license/doc/reference/trd-legal.md)

**Comment by jrvanwhy** [comment]
> We'll probably end up with an exceptions list in the tool. For example, if someone copies a file from another project without making any changes to it, then it should be committed without a Tock project notice. In that case, we'll need to list the file as an exception in the tool. When we eventually make our own contributions to that file, we will then add a Tock project notice and remove the license exception.

I just finished typing up the [notes](https://github.com/tock/tock/pull/3326) from today's core WG call. Having listened to the discussion again, I've changed my mind slightly on this.

If we copy a file from another (Apache 2.0 OR MIT) project into Tock, I think it is totally fine to add something like:

```
// Licensed under the Apache License, Version 2.0 or the MIT License.
// SPDX-License-Identifier: Apache-2.0 OR MIT
```

at the top of the file. It's the copyright line that we shouldn't add if we didn't contribute to the file.

However, I do think we should allow some variation in the format if we use a file from another project. For example, if we copy in a file from another project with the following header:

```
// Copyright ExampleCo
// Licensed under the Apache License, Version 2.0 or the MIT license, at your option.
// SPDX-License-Identifier: Apache-2.0 OR MIT
```

I think we should leave it intact, rather than requiring us to change it to:

```
// Licensed under the Apache License, Version 2.0 or the MIT License.
// SPDX-License-Identifier: Apache-2.0 OR MIT
// Copyright ExampleCo
// Licensed under the Apache License, Version 2.0 or the MIT license, at your option.
// SPDX-License-Identifier: Apache-2.0 OR MIT
```

and the enforcement tool should be flexible enough to make that easy.

**Review by brghena — APPROVED** [review]

**Review by alevy — DISMISSED** [review]

File: doc/reference/trd-legal.md
Comment (jrvanwhy) [inline doc/reference/trd-legal.md:75]: What about files that do not support comments? Markdown comes to mind -- I suspect we want to omit license notices in our Markdown documents.

**Review by jrvanwhy — COMMENTED** [review]
Based on my reading of the document, I think the enforcement tool should:

1. Verify the license information is present near the top of each file and formatted according to this policy.
1. NOT look for license information elsewhere in the file, and ignore any that is present.

With this implementation, we could copy code from other projects with license/copyright information in the middle of the file without adding an exception in the tool. The drawback is that it would not catch errant license headers in the middle of new contributions.

What do you think?

File: doc/reference/trd-legal.md
Comment (jrvanwhy) [reply doc/reference/trd-legal.md:75]: I suspect that is fine.

Comment (hudson-ayers) [reply doc/reference/trd-legal.md:75]: It sounds like you two have different answers to the question "is markdown code?"

Comment (jrvanwhy) [reply doc/reference/trd-legal.md:75]: I'm not sure that either "Markdown vs not-Markdown" or "code vs not-code" is the right place to make the distinction. How about requiring headers in all file formats that support comments? I don't want someone to think that images have to have license info embedded in their EXIF tags, so maybe all *textual* file formats that support comments.

I don't particularly care where we draw the line, I just think we need to be unambiguous.

Comment (alevy) [reply doc/reference/trd-legal.md:75]: Sure, that seems fine to me.

File: doc/reference/trd-legal.md
Comment (lschuermann) [inline doc/reference/trd-legal.md:92]: We may want to use this "loophole" to avoid having the license specification be the first lines of a file which also includes a module-level documentation comment. In my (rather subjective) opinion, something like the following actually looks rather odd to people who expect module documentation to always be at the top of a file (which is motivated by Rust not allowing any actual code before such a parent-module doc comment). This would be a generous interpretation of _technical or other reasons_:

```
// Licensed under the Apache License, Version 2.0 or the MIT License.
// SPDX-License-Identifier: Apache-2.0 OR MIT
// Copyright Jane Doe <jdoe@example.org> 2019,2022

//! Provides userspace with access to a serial interface.
//!
//! Setup
//! -----
//!
//! You need a device that provides the `hil::uart::UART` trait.
```

I am aware this conflicts with the example provided below. Not trying to start a flamewar about code style here, just my 2ct.

**Review by lschuermann — APPROVED** [review]
This looks pretty good to me.

File: doc/reference/trd-legal.md
Comment (jrvanwhy) [reply doc/reference/trd-legal.md:92]: I think I interpret "necessary" different than you do. If we want the headers to go below module documentation, then we should state that explicitly.

Comment (lschuermann) [reply doc/reference/trd-legal.md:92]: Oh, yeah, I agree with you. My comment is a convoluted way to ask the question "to me this looks weird, should we do this instead?". If we come to the conclusion that license specs should indeed follow doc-comments or even be part of them (at the end, which I've seen throughout the ecosystem), we might want to change the TRD.

File: doc/reference/trd-legal.md
Comment (alistair23) [inline doc/reference/trd-legal.md:50]: This doesn't seem right.

What about tools? What if there are tools that run on the host. For example scripts that we might use to run tests, checks or flash binaries.

Also, what about newlib which is used by libtock-c? The binaries are [committed to a Tock project](https://github.com/tock/libtock-c/tree/master/newlib/cortex-m/v7-m)

File: doc/reference/trd-legal.md
Comment (hudson-ayers) [reply doc/reference/trd-legal.md]: I realize this was already updated from MUST to SHOULD, but I think we should still say "at least one". For example, [PEP 8 requires top-level functions and classes be surrounded by two blank lines](https://peps.python.org/pep-0008/#blank-lines), so most Python files without other comments at the top of the file would want two spaces after this information.

File: doc/reference/trd-legal.md
Comment (hudson-ayers) [reply doc/reference/trd-legal.md:67]: Looking...*very* far into the future, one reason for a non-author to remove a copyright would be if the copyright has expired. 

Regardless, I don't think we should interpret the use of `SHOULD NOT` as implying `MAY`, it is just a hedge that we may not have considered every possible case. Obviously any removal of a copyright notice is something that should get attention in a code review.

File: doc/reference/trd-legal.md
Comment (hudson-ayers) [reply doc/reference/trd-legal.md:50]: I do think it is a good question whether this should actually apply to all Tock project code -- for example, I don't think we want to go and add copyright headers to every file in the tock-teensy repo. Similarly, a tool like tockloader is a published artifact, and even most corporate users will not be depending on any of its source files directly -- they just might invoke the tool at some point during the process of building/flashing.

@jrvanwhy for Google's purposes, do you know what subset of Tock project repositories need to have license headers? For example, would just the kernel be sufficient? Or the kernel + libtock-c? I think we should limit the scope of this as much as possible.

File: doc/reference/trd-legal.md
Comment (alevy) [reply doc/reference/trd-legal.md]: Resolution from call: "SHOULD have *at least one*"

File: doc/reference/trd-legal.md
Comment (alevy) [reply doc/reference/trd-legal.md:75]: Conclusion: Textual files that allow comments

(e.g. Markdown files don't support comments, so we don't expect it)

File: doc/reference/trd-legal.md
Comment (alevy) [reply doc/reference/trd-legal.md:92]: Conclusion: redundant since we're going to be more liberal with code imported from third-parties. Normal practice would be license header above module docs.

File: doc/reference/trd-legal.md
Comment (alevy) [reply doc/reference/trd-legal.md:50]: Conclusion 1: It's a lot easier if everything in a repository is under the same/compatible licenses. Let's keep it to that for Tock.

Conclusion 2: For other repositories, let's punt with this document. I.e. change the language to refer _specifically_ to _this_ Tock repository, not the Tock "project" as a whole.

**Comment by alevy** [comment]
Additional conclusions:
- The expectations for formatting are about the Tock notices, other notices don't have to follow these requirements.
- If notices different enough to modify then instead create a new one above it, but should be flexible to accept headers that are close enough.

**Comment by bradjc** [comment]
I fixed the header formatting.

**Comment by jrvanwhy** [comment]
> Overall this approach looks good to me. Curious to hear what @jrvanwhy in particular thinks about not requiring the "Copyright Tock Contributors" line on all files without a different copyright assignment

I just noticed I never answered this question. Personally, I'm not a fan of making the copyright notice optional. Making the copyright notice optional will tend to overrepresent the contributions of some contributors relative to others.

For example, suppose contributors A, B, C, D, and E work together on a file (all contributing significantly), but the only contributor who wants to add a copyright notice is A. In that case, the license header will become:

```
# Licensed under the Apache License, Version 2.0 or the MIT License.
# SPDX-License-Identifier: Apache-2.0 OR MIT
# Copyright A 2022.
```

Someone who is *not* from the Tock project who stumbles across this file will see it as primarily A's work, rather than a community project.

Requiring the "Copyright Tock Contributors" line would change that to:

```
# Licensed under the Apache License, Version 2.0 or the MIT License.
# SPDX-License-Identifier: Apache-2.0 OR MIT
# Copyright Tock Contributors 2022.
# Copyright A 2022.
```

which is a more accurate representation of the code's authorship.

**Comment by lschuermann** [comment]
I share the spirit of @jrvanwhy's comment. Yet, I'm unsure whether requiring such a line might not introduce some issues: obviously we'd not add such a line on a compatibly-licensed, vendored file. But even when a file is added e.g. by its original author to Tock, I think it might be weird to force this developer to attribute the wider "Tock contributors" despite the file not having received any contributions other than the original author's (yet). I think having a "Copyright Tock Contributors" line in each file as a convention is good, and I would adapt that, but I'm wary about enforcing it. I presume every contributor would be able to decide on their own whether to attribute just themselves, the "Tock contributors", or both.

**Comment by jrvanwhy** [comment]
We could require the "Copyright Tock Contributors" line only in cases where non-trivial contributions are made by contributors who do not have their own copyright lines. In practice, I think this would result in most files carrying only the "Copyright Tock Contributors" line (because most of y'all seem to prefer that over per-person copyrights), and a handful of files having 1 or 2 specific contributors named.

The downside is that "non-trivial contributions" is a bit vague and therefore this cannot be automatically enforced, but I don't expect that will be an issue in practice.

**Comment by hudson-ayers** [comment]
> We could require the "Copyright Tock Contributors" line only in cases where non-trivial contributions are made by contributors who do not have their own copyright lines. In practice, I think this would result in most files carrying only the "Copyright Tock Contributors" line (because most of y'all seem to prefer that over per-person copyrights), and a handful of files having 1 or 2 specific contributors named.
> 
> The downside is that "non-trivial contributions" is a bit vague and therefore this cannot be automatically enforced, but I don't expect that will be an issue in practice.

I disagree that this will not be an issue in practice. What will happen without automatic enforcement is there will be tons of files with "Copyright FIRST_AUTHOR" at the top that have in fact been modified by dozens of other people. For example, Phil used to often add an `Author: Philip Levis` line to the top of files he created. If you grep for that in the Tock repo, you will find dozens of files with just "Author: Philip Levis" or maybe "Author: Philip Levis, Leon Schuermann" where looking at the git history will reveal many other authors with significant contributions.

I think it is much easier in practice to just require "Copyright Tock Contributors". If an individual author wishes to add an additional copyright for just their own name they are free to do so.

**Comment by alevy** [comment]
I'm going to convert this from a Draft PR to a real PR. I _believe_ the remaining issue is just whether to require a copyright line or make it optional. Please comment if there is more missing still.

**Review by alevy — DISMISSED** [review]

File: doc/reference/trd-legal.md
Comment (jrvanwhy) [inline doc/reference/trd-legal.md:125]: The Examples section needs to be updated to indicate that a copyright line is necessary.

We seem to have come to the conclusion (see [Hudson's comment](https://github.com/tock/tock/pull/3318#issuecomment-1324225961)) that we should require "Copyright Tock Contributors" to be one of the copyright lines, or at least make it standard, so it should probably be part of the examples as well.

Comment (lschuermann) [reply doc/reference/trd-legal.md:125]: At the risk of repeating what I said above, I’d vouch for making it standard, as I foresee issues with requiring this line for files added without modification by a “Tock Contributor”. Otherwise I agree that this section should be updated.

File: doc/reference/trd-legal.md
Comment (jrvanwhy) [inline doc/reference/trd-legal.md:77]: From the 2023-01-13 core wg call: This line needs an exception for files that Tock contributors have not made contributions for (e.g. vendored files).

Comment (jrvanwhy) [reply doc/reference/trd-legal.md]: Merged #3377 which adds an exception for vendored and other imported files.

File: doc/reference/trd-legal.md
Comment (jrvanwhy) [reply doc/reference/trd-legal.md:126]: Merged #3377 which updated the examples. `Copyright Tock Contributors` is part of the examples but is not required in the text.

File: doc/reference/trd-legal.md
Comment (jrvanwhy) [reply doc/reference/trd-legal.md]: This PR has been updated to only apply to the `tock` repository. We expect to discuss applying this policy to other repositories after this PR has been merged.

File: doc/reference/trd-legal.md
Comment (jrvanwhy) [reply doc/reference/trd-legal.md]: I changed it to "at least one".

**Review by hudson-ayers — APPROVED** [review]

**Review by lschuermann — APPROVED** [review]

**Review by brghena — APPROVED** [review]

**Review by alevy — APPROVED** [review]

**Pull request merged**
